File: ros2_efficiency_test/src/control_node.cpp
#include <geometry_msgs/msg/twist.hpp>
#include <message_filters/subscriber.h>
#include <message_filters/time_synchronizer.h>
#include <nav_msgs/msg/odometry.hpp>
#include <nav_msgs/msg/path.hpp>
#include <rclcpp/rclcpp.hpp>

class ControlNode : public rclcpp::Node {
public:
  ControlNode() : Node("control_node") {
    path_sub_ = this->create_subscription<nav_msgs::msg::Path>(
        "path", 10,
        std::bind(&ControlNode::path_callback, this, std::placeholders::_1));

    odom_sub_ = this->create_subscription<nav_msgs::msg::Odometry>(
        "odometry/rate_200hz", 10,
        std::bind(&ControlNode::odom_callback, this, std::placeholders::_1));

    cmd_pub_ = this->create_publisher<geometry_msgs::msg::Twist>("cmd_vel", 10);

    timer_ = this->create_wall_timer(
        std::chrono::milliseconds(5),
        std::bind(&ControlNode::control_callback, this));
  }

private:
  void path_callback(const nav_msgs::msg::Path::SharedPtr msg) {
    latest_path_ = msg;
  }

  void odom_callback(const nav_msgs::msg::Odometry::SharedPtr msg) {
    latest_odom_ = msg;
  }

  void control_callback() {
    if (!latest_path_ || !latest_odom_)
      return;

    auto cmd_msg = geometry_msgs::msg::Twist();
    cmd_msg.linear.x = 0.5;
    cmd_msg.angular.z = 0.1;
    cmd_pub_->publish(cmd_msg);
  }

  rclcpp::Subscription<nav_msgs::msg::Path>::SharedPtr path_sub_;
  rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr odom_sub_;
  rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr cmd_pub_;
  rclcpp::TimerBase::SharedPtr timer_;

  nav_msgs::msg::Path::SharedPtr latest_path_;
  nav_msgs::msg::Odometry::SharedPtr latest_odom_;
};

int main(int argc, char *argv[]) {
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<ControlNode>());
  rclcpp::shutdown();
  return 0;
}

File: ros2_ws/src/ros2_efficiency_test_py/src/control_node.py
#!/usr/bin/env python3
import os

import rclpy
from rclpy.experimental.events_executor import EventsExecutor
from rclpy.node import Node
from geometry_msgs.msg import Twist
from nav_msgs.msg import Odometry, Path


class ControlNode(Node):
    def __init__(self):
        super().__init__('control_node')

        self.path_sub_ = self.create_subscription(
            Path, 'path', self.path_callback, 10)
        self.odom_sub_ = self.create_subscription(
            Odometry, 'odometry/rate_200hz', self.odom_callback, 10)

        self.cmd_pub_ = self.create_publisher(Twist, 'cmd_vel', 10)

        self.timer_ = self.create_timer(0.005, self.control_callback)  # 5ms = 200Hz

        self.latest_path_ = None
        self.latest_odom_ = None

    def path_callback(self, msg):
        self.latest_path_ = msg

    def odom_callback(self, msg):
        self.latest_odom_ = msg

    def control_callback(self):
        if self.latest_path_ is None or self.latest_odom_ is None:
            return

        cmd_msg = Twist()
        cmd_msg.linear.x = 0.5
        cmd_msg.angular.z = 0.1
        self.cmd_pub_.publish(cmd_msg)


def main(args=None):
    rclpy.init(args=args)
    node = ControlNode()
    executor = None

    try:
        if os.environ.get('EVENT_EXECUTOR') == '1':
            executor = EventsExecutor()
            executor.add_node(node)
            executor.spin()
        else:
            rclpy.spin(node)
    finally:
        if executor is not None:
            executor.shutdown()
        node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()

if __name__ == '__main__':
    main()
